File: src/solutions/day02/day02.py
from typing import List, Tuple
BASE_PATH = "../../inputs/day02"

def generic_solution(numbers: List[int], window_size=1) -> int:
    return sum([int(numbers[i] < numbers[i+window_size]) for i in range(len(numbers)-window_size)])

def solution_1(instructions) -> Tuple[int, int]:
    depth = horizontal = 0
    for instruction, value in instructions:
        if instruction == "forward":
            horizontal += value
        if instruction == "up":
            depth -= value
        if instruction == "down":
            depth += value
    return (horizontal, depth)

def solution_2(instructions) -> Tuple[int, int]:
    depth = horizontal = aim = 0
    for instruction, value in instructions:
        if instruction == "forward":
            horizontal += value
            depth += aim * value
        if instruction == "up":
            aim -= value
        if instruction == "down":
            aim += value
    return (horizontal, depth)



for file_name in ["test.txt", "sol.txt"]:
    with open(f"{BASE_PATH}/{file_name}") as f:
        lines = [[i, int(c)] for i, c in [l.strip().split() for l in f if l.strip()]]


    print(f"Solutions for {file_name}:")
    sol1 = solution_1(lines)
    print(f"Part 1: {sol1[0] * sol1[1]}")
    sol2= solution_2(lines)
    print(f"Part 2: {sol2[0] * sol2[1]}")

File: src/solutions/day02/day02.rs
fn solution_1(file_content: &str) -> usize {
    let mut depth = 0;
    let mut horizontal = 0;
    for line in file_content.lines() {
        let split_line: Vec<&str> = line.trim().split(" ").collect();
        let instruction = split_line[0];
        let value = split_line[1].parse::<usize>().expect("Cannot parse value");

        match instruction {
            "forward" => horizontal += value,
            "down" => {
                depth += value;
            }
            "up" => {
                depth -= value;
            }
            _ => {
                println!("Unknown operation!!")
            }
        }
    }
    return depth * horizontal;
}

fn solution_2(file_content: &str) -> usize {
    let mut depth = 0;
    let mut horizontal = 0;
    let mut aim = 0;
    for line in file_content.lines() {
        let split_line: Vec<&str> = line.trim().split(" ").collect();
        let instruction = split_line[0];
        let value = split_line[1].parse::<usize>().expect("Cannot parse value");

        match instruction {
            "forward" => {
                horizontal += value;
                depth += aim * value;
            }
            "down" => {
                aim += value;
            }
            "up" => {
                aim -= value;
            }
            _ => {
                println!("Unknown operation!!")
            }
        }
    }
    return depth * horizontal;
}

pub fn day02() {
    println!("Solutions for test.txt:");
    let content = include_str!("../../inputs/day02/test.txt");
    println!("Part 1 {}", solution_1(content));
    println!("Part 2 {}", solution_2(content));

    println!("Solutions for sol.txt:");
    let content = include_str!("../../inputs/day02/sol.txt");
    println!("Part 1 {}", solution_1(content));
    println!("Part 2 {}", solution_2(content));
}
